src/language-services/analysis.ts

```diff
@@ -165,82 +162,108 @@ class DocumentAnalysis implements Analysis {
-    const maybeLineTokens: ReadonlyArray<PQP.Language.LineToken> | undefined = maybeLineTokensAt(document, position);
+    private maybeLineTokensAt(): ReadonlyArray<PQP.Language.LineToken> | undefined {
+        const lexResult: PQP.Lexer.State = this.getLexerState();
+        const maybeLine: PQP.Lexer.TLine | undefined = lexResult.lines[this.position.line];
 
-    if (maybeLineTokens === undefined) {
-        return undefined;
+        return maybeLine !== undefined ? maybeLine.tokens : undefined;
     }
 
-    const lineTokens: ReadonlyArray<PQP.Language.LineToken> = maybeLineTokens;
+    private maybeTokenAt(): PQP.Language.LineToken | undefined {
+        const maybeLineTokens: ReadonlyArray<PQP.Language.LineToken> | undefined = this.maybeLineTokensAt();
 
-    for (const token of lineTokens) {
-        if (token.positionStart <= position.character && token.positionEnd >= position.character) {
-            return token;
+        if (maybeLineTokens === undefined) {
+            return undefined;
         }
-    }
 
-    // Token wasn't found - check for special case where current position is a trailing "." on an identifier
-    const currentRange: Range = {
-        start: {
-            line: position.line,
-            character: position.character - 1,
-        },
-        end: position,
-    };
+        const lineTokens: ReadonlyArray<PQP.Language.LineToken> = maybeLineTokens;
 
-    if (document.getText(currentRange) === ".") {
         for (const token of lineTokens) {
-            if (token.positionStart <= position.character - 1 && token.positionEnd >= position.character - 1) {
-                if (token.kind === PQP.Language.LineTokenKind.Identifier) {
-                    // Use this token with an adjusted position
-                    return {
-                        data: `${token.data}.`,
-                        kind: token.kind,
-                        positionStart: token.positionStart,
-                        positionEnd: token.positionEnd + 1,
-                    };
+            if (token.positionStart <= this.position.character && token.positionEnd >= this.position.character) {
+                return token;
+            }
+        }
+
+        // TODO: is this still needed with the latest parser?
+        // Token wasn't found - check for special case where current position is a trailing "." on an identifier
+        const currentRange: Range = {
+            start: {
+                line: this.position.line,
+                character: this.position.character - 1,
+            },
+            end: this.position,
+        };
+
+        if (this.getText(currentRange) === ".") {
+            for (const token of lineTokens) {
+                if (
+                    token.positionStart <= this.position.character - 1 &&
+                    token.positionEnd >= this.position.character - 1
+                ) {
+                    if (token.kind === PQP.Language.LineTokenKind.Identifier) {
```

Comment:
Fold the second if statement into the first. Make the tokenKind check be the first condition for a tiny optimization.